lib/Echidna/Campaign.hs

```diff
@@ -241,20 +244,91 @@ addToCorpus :: (MonadState s m, Has Campaign s) => Integer -> [(Tx, (VMResult, I
 
-seqMutators :: (MonadRandom m) => MutationConsts -> m (Int -> Corpus -> [Tx] -> m [Tx])
-seqMutators (c1, c2, c3) = fromList 
-                            [(cnm      , fromInteger c1),
-                             (mut False, fromInteger c2),
-                             (mut True , fromInteger c3)]
-  where -- Use the generated random transactions
-        cnm _ _          = return
-        mut flp ql ctxs gtxs = do
-          let somePercent = if (fst . DS.findMax) ctxs > 1    -- if the corpus already contains new elements 
-                             then 1 + (DS.size ctxs `div` 20) -- then take 5% of its size
-                             else DS.size ctxs                -- otherwise, take all of it
-          rtxs <- fromList $ map (\(i, txs) -> (txs, fromInteger i)) $ take somePercent $ DS.toDescList ctxs
-          k <- getRandomR (0, length rtxs - 1)
-          return . take ql $ if flp then take k gtxs ++ rtxs else take k rtxs ++ gtxs
+data TxsMutation = TxIdentity
+                 | Shrinking
+                 | Mutation
+                 | Expansion
+                 | Swapping
+                 | Deletion
+
+data CorpusMutation = Skip
+                    | RandomAppend TxsMutation 
+                    | RandomPrepend TxsMutation
+                    | RandomSplice
+                    | RandomInterleave
+
+selectAndMutate :: MonadRandom m 
+                => ([Tx] -> m [Tx]) -> Corpus -> m [Tx]
+selectAndMutate f ctxs = do
+  let somePercent = if (fst . DS.findMax) ctxs > 1   -- if the corpus already contains new elements 
+                    then 1 + (DS.size ctxs `div` 20) -- then take 5% of its size
+                    else DS.size ctxs                -- otherwise, take all of it
+  rtxs <- fromList $ map (\(i, txs) -> (txs, fromInteger i)) $ take somePercent $ DS.toDescList ctxs
+  k <- getRandomR (0, length rtxs - 1)
+  f $ take k rtxs
+
+
+selectAndCombine ::  MonadRandom m
+                 => ([Tx] -> [Tx] -> m [Tx]) -> Int -> Corpus -> [Tx] -> m [Tx]
+selectAndCombine f ql ctxs gtxs = do
+          rtxs1 <- fromList $ map (\(i, txs) -> (txs, fromInteger i)) $ DS.toDescList ctxs
+          rtxs2 <- fromList $ map (\(i, txs) -> (txs, fromInteger i)) $ DS.toDescList ctxs 
+          txs <- f rtxs1 rtxs2
+          return . take ql $ txs ++ gtxs
+
+getCorpusMutation :: (MonadThrow m, MonadRandom m, Has GenDict x, MonadState x m) 
+                  => CorpusMutation -> Int -> Corpus -> [Tx] -> m [Tx]
+getCorpusMutation Skip = \_ _ -> return
+getCorpusMutation (RandomAppend m) = 
+  case m of
+    TxIdentity -> mut return
+    Shrinking  -> mut (mapM shrinkTx)
+    Mutation   -> mut (mapM mutateTx)
+    Expansion  -> mut expandRandList
+    Swapping   -> mut swapRandList
+    Deletion   -> mut deleteRandList
+ where mut f ql ctxs gtxs = do
+          rtxs' <- selectAndMutate f ctxs
+          return . take ql $ rtxs' ++ gtxs  
+
+getCorpusMutation (RandomPrepend m) = 
+  case m of
+    TxIdentity -> mut return
```

Comment:
there is double skip happening, is this intentional?

Reply:
Same comment as `RandomAppend`.